### PR TITLE
[SYCL][HIP] Keep track of only shared USM allocations and prefetch only for those

### PR DESCRIPTION
**Pull request by hdelan**
https://github.com/intel/llvm/pull/10430 caused a perf regression. This does the same prefetching but only for shared USM allocations, not for all USM allocations. This should revert some of the perf regression for non shared USM. Patch to follow to minimize perf hit unless using problematic instructions.

### CODE REVIEW

File: libclc/amdgcn-amdhsa/libspirv/atomic/atomic_sub.cl [hunk 1]
```diff
@@ -18,7 +18,7 @@
           enum MemorySemanticsMask semantics, TYPE val) {                                                                     \
```
Comment (jchlanda) [inline libclc/amdgcn-amdhsa/libspirv/atomic/atomic_sub.cl:18]: The changes to this file should be exported to a separate PR, they also deserve an explanation for the sign flipping.

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool runImpl(Module &M) {
+  if (moduleHasAtomicXor(M)) {
```
Comment (jchlanda) [inline llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:41]: I'd flip the condition, so you do en early exit here and save yourself one level of indentation.

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        if (AtomicInst->getOperation() == AtomicRMWInst::Xor) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool runImpl(Module &M) {
```
Comment (jchlanda) [inline llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:40]: I'd probably get rid of this and fold it in to `run` it doesn't encapsulate any information, hence a bit useless.

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  if (moduleHasAtomicXor(M)) {
+    LLVMContext &Ctx = M.getContext();
+    new GlobalVariable(M, Type::getInt1Ty(Ctx), true,
+                       GlobalValue::InternalLinkage,
+                       Constant::getAllOnesValue(Type::getInt1Ty(Ctx)),
+                       NEW_GLOBAL_NAME); // FIXME: this seems wrong and bad, is
```
Comment (jchlanda) [inline llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:46]: Maybe this would be better: https://llvm.org/doxygen/Module_8cpp_source.html#l00205
it checks if the global exists in a first place?

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#include "llvm/IR/Instructions.h"
+
+using namespace llvm;
+
+#define NEW_GLOBAL_NAME "HipAtomicXorModuleNeedsPrefetch"
+
+namespace {
+
+bool moduleHasAtomicXor(Module &M) {
+  for (auto &F : M) {
```
Comment (jchlanda) [inline llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:28]: I'd need to look up the coding standards, but this might qualify for no braces.

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
+//===- AMDGPUAddGlobalForAtomicXor.cpp ------------------------------------===//
```
Comment (jchlanda) [inline llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:1]: I'm not sure if this pass should really live in the AMDGPU backend, it is so closely coupled with SYCL runtime that one could argue it should be contained within [SYCLLowerIR](https://github.com/intel/llvm/tree/sycl/llvm/lib/SYCLLowerIR).

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp (hunk 4 above)
Comment (hdelan) [reply llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:46]: Spot on

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp (hunk 2 above)
Comment (hdelan) [reply llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:41]: Nice thanks

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp (hunk 3 above)
Comment (hdelan) [reply llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:40]: Yep good idea

File: libclc/amdgcn-amdhsa/libspirv/atomic/atomic_sub.cl (hunk 1 above)
Comment (hdelan) [reply libclc/amdgcn-amdhsa/libspirv/atomic/atomic_sub.cl:18]: Will do

File: llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp (hunk 5 above)
Comment (hdelan) [reply llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp:28]: I think you are correct

Comment (jchlanda) [reply llvm/lib/Target/AMDGPU/AMDGPUAddGlobalForAtomicXor.cpp]: For completeness: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
